server/resp/src/main/java/org/infinispan/server/resp/SubscriberHandler.java

```diff
@@ -0,0 +1,166 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletionStage;
+
+import org.infinispan.commons.logging.LogFactory;
+import org.infinispan.commons.marshall.WrappedByteArray;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryCreated;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryModified;
+import org.infinispan.notifications.cachelistener.event.CacheEntryEvent;
+import org.infinispan.server.core.logging.Log;
+import org.infinispan.util.concurrent.CompletableFutures;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.util.CharsetUtil;
+
+public class SubscriberHandler implements RespRequestHandler {
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass(), Log.class);
+   public static final byte[] PREFIX_CHANNEL_BYTES = new byte[]{-114, 16, 78, -3, 127};
+   private final Resp3Handler handler;
+   private final RespServer respServer;
+
+   public SubscriberHandler(RespServer respServer) {
+      this.respServer = respServer;
+      this.handler = respServer.getHandler();
+   }
+
+   public static byte[] keyToChannel(byte[] keyBytes) {
+      byte[] result = new byte[keyBytes.length + PREFIX_CHANNEL_BYTES.length];
+      System.arraycopy(PREFIX_CHANNEL_BYTES, 0, result, 0, PREFIX_CHANNEL_BYTES.length);
+      System.arraycopy(keyBytes, 0, result, PREFIX_CHANNEL_BYTES.length, keyBytes.length);
+      return result;
+   }
+
+   public static byte[] channelToKey(byte[] channelBytes) {
+      return Arrays.copyOfRange(channelBytes, PREFIX_CHANNEL_BYTES.length, channelBytes.length);
+   }
+
+   @Listener(clustered = true)
+   static class PubSubListener {
+      private final Channel channel;
+
+      PubSubListener(Channel channel) {
+         this.channel = channel;
+      }
+
+      @CacheEntryCreated
+      @CacheEntryModified
+      public CompletionStage<Void> onEvent(CacheEntryEvent<byte[], byte[]> entryEvent) {
+         byte[] key = channelToKey(entryEvent.getKey());
+         byte[] value = entryEvent.getValue();
+         if (key.length > 0 && value != null && value.length > 0) {
+            ByteBuf byteBuf = channel.alloc().buffer(2 + 2
+                  + 2 + 7 + 1 + (int) Math.log10(key.length) + 1 + 2
+                  + key.length + 2 + 1 + (int) Math.log10(value.length) + 1 + 2 + value.length + 2);
+            byteBuf.writeCharSequence("*3\r\n$7\r\nmessage\r\n$" + key.length + "\r\n", CharsetUtil.UTF_8);
+            byteBuf.writeBytes(key);
+            byteBuf.writeCharSequence("\r\n$" + value.length + "\r\n", CharsetUtil.UTF_8);
+            byteBuf.writeBytes(value);
+            byteBuf.writeCharSequence("\r\n", CharsetUtil.UTF_8);
+            // TODO: add some back pressure? - something like ClientListenerRegistry?
+            channel.writeAndFlush(byteBuf);
```

Comment:
No, what this needs to do is check if the channel is writeable and if it is just write to otherwise offload it to be done later. This needs to be similar to https://github.com/infinispan/infinispan/blob/main/server/hotrod/src/main/java/org/infinispan/server/hotrod/ClientListenerRegistry.java#L387. I will log a JIRA to take care of this after this PR though.